File: java/src/Person.java
/**
 * Created by zixuan on 2016/9/1.
 */

import java.lang.Math;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.logging.Level;

public class Person {
    public Person(ArrayList<Signature> refSigs)
    {
        // Utils.logger.log(Level.INFO, "Reference signature count: {0}", refSigs.size());
        this.refSigs = refSigs;
        // Select template signature
        this.SelectTemplate();
        // Calculate base distance
        this.baseValueMap = new HashMap<>();
        this.CalcBaseDis();
    }


    /*
    @Description: Select template signature from reference signatures.
     */
    public void SelectTemplate()
    {
        // Utils.logger.log(Level.INFO, "Selecting template signature");
        ArrayList<Double> refDis = new ArrayList<>();
        for (int i = 0; i < this.GetRefCount(); ++i)
        {
            double dis = 0.0;
            for (int j = 0; j < this.GetRefCount(); ++j)
            {
                if (i == j)
                    continue;
                ArrayList<Double> comDisList = new ArrayList<>();
                for (String com : Config.SigComList)
                {
                    ArrayList<Double> signal1 = refSigs.get(i).GetCom(com);
                    ArrayList<Double> signal2 = refSigs.get(j).GetCom(com);
                    // comDisList.add(NaiveDTW(signal1, signal2, Config.Penalization.get(com), Config.Threshold.get(com)));
                    comDisList.add(NaiveDTW(refSigs.get(i).GetCom(com), refSigs.get(j).GetCom(com), Config.Penalization.get(com), Config.Threshold.get(com)));
                }
                dis += Utils.GetMeanValue(comDisList);
            }
            refDis.add(dis);
        }
        this.templateIndex = Utils.GetMinIndex(refDis);
        this.templateSig = refSigs.get(this.templateIndex);
        // Utils.logger.log(Level.INFO, "Template index: {0}.", new Object[]{templateIndex});
    }

    /*
    @Description: Calculate the base value in signal component of signatures
     */
    public void CalcBaseDis()
    {
        // Utils.logger.log(Level.INFO, "calculating base distance");
        for (int k = 0; k < Config.SigComList.size(); ++k)
        {
            String com = Config.SigComList.get(k);
            ArrayList<Double> templateComList = new ArrayList<>();
            ArrayList<Double> maxComList = new ArrayList<>();
            ArrayList<Double> minComList = new ArrayList<>();
            ArrayList<Double> avgComList = new ArrayList<>();

            for (int i = 0; i < this.GetRefCount(); ++i)
            {
                if (i == this.templateIndex)
                    continue;
                ArrayList<Double> comi = this.refSigs.get(i).GetCom(com);
                Double templateComDis = NaiveDTW(comi, this.templateSig.GetCom(com),
                        Config.Penalization.get(com), Config.Threshold.get(com));
                templateComList.add(templateComDis);
                ArrayList<Double> comDisList = new ArrayList<>();
                for (int j = 0; j < this.GetRefCount(); ++j)
                {
                    if (i == j)
                        continue;
                    ArrayList<Double> comj = this.refSigs.get(j).GetCom(com);
                    comDisList.add(NaiveDTW(comi, comj, Config.Penalization.get(com), Config.Threshold.get(com)));
                }
                maxComList.add(Utils.GetMaxValue(comDisList));
                minComList.add(Utils.GetMinValue(comDisList));
                avgComList.add(Utils.GetMeanValue(comDisList));
            }
            if (Config.FeatureType.get(com).indexOf(Config.Template) != -1)
                this.baseValueMap.put(Config.Template+com, Utils.GetMeanValue(templateComList));
            if (Config.FeatureType.get(com).indexOf(Config.Max) != -1)
                this.baseValueMap.put(Config.Max+com, Utils.GetMeanValue(maxComList));
            if (Config.FeatureType.get(com).indexOf(Config.Min) != -1)
                this.baseValueMap.put(Config.Min+com, Utils.GetMeanValue(minComList));
            if (Config.FeatureType.get(com).indexOf(Config.Avg) != -1)
                this.baseValueMap.put(Config.Avg+com, Utils.GetMeanValue(avgComList));
            if (Config.FeatureType.get(com).indexOf(Config.Median) != -1)
                this.baseValueMap.put(Config.Median+com, Utils.GetMedianValue(avgComList));
            /*
            Utils.logger.log(Level.INFO, "Calculating signal: {0} {1} {2} {3} {4}",
                    new Object[]{com,
                    this.baseValueMap.get(Config.Template+com),
                    this.baseValueMap.get(Config.Max+com),
                    this.baseValueMap.get(Config.Min+com),
                    this.baseValueMap.get(Config.Avg+com)});
            */
        }
    }

    /*
    @Description: For given signature, calculate vector[] with normalization.
     */
    public ArrayList<Double> CalcDis(Signature signature)
    {
        ArrayList<Double> featureList = new ArrayList<>();
        for (String com : Config.SigComList)
        {
            ArrayList<Double> comSig = signature.GetCom(com);
            ArrayList<Double> comTemplate = this.templateSig.GetCom(com);
            Double templateComDis = NaiveDTW(comSig, comTemplate, Config.Penalization.get(com), Config.Threshold.get(com));
            ArrayList<Double> comDisList = new ArrayList<>();
            for (int i = 0; i < this.GetRefCount(); ++i)
            {
                ArrayList<Double> comI = this.refSigs.get(i).GetCom(com);
                Double dis = NaiveDTW(comSig, comI, Config.Penalization.get(com), Config.Threshold.get(com));
                comDisList.add(dis);
            }
            Double maxComDis = Utils.GetMaxValue(comDisList);
            Double minComDis = Utils.GetMinValue(comDisList);
            Double avgComDis = Utils.GetMeanValue(comDisList);
            Double medComDis = Utils.GetMedianValue(comDisList);
            if (Config.FeatureType.get(com).indexOf(Config.Template) != -1)
                featureList.add(templateComDis / this.baseValueMap.get(Config.Template+com));
            if (Config.FeatureType.get(com).indexOf(Config.Max) != -1)
                featureList.add(maxComDis / this.baseValueMap.get(Config.Max+com));
            if (Config.FeatureType.get(com).indexOf(Config.Min) != -1)
                featureList.add(minComDis / this.baseValueMap.get(Config.Min+com));
            if (Config.FeatureType.get(com).indexOf(Config.Avg) != -1)
                featureList.add(avgComDis / this.baseValueMap.get(Config.Avg+com));
            if (Config.FeatureType.get(com).indexOf(Config.Median) != -1)
                featureList.add(medComDis / this.baseValueMap.get(Config.Median+com));
        }
        return featureList;
    }

    public int GetRefCount()
    {
        return this.refSigs.size();
    }

    public Double NaiveDTW(ArrayList<Double> A, ArrayList<Double> B, Double p, Double t)
    {
        int size1 = A.size();
        int size2 = B.size();
        Double distance[][] = new Double[size1][size2];
        distance[0][0] = Math.abs(A.get(0) - B.get(0));
        for (int i = 1; i < size1; ++i)
            distance[i][0] = distance[i-1][0] + Math.abs(A.get(i) - B.get(0));
        for (int i = 1; i < size2; ++i)
            distance[0][i] = distance[0][i-1] + Math.abs(A.get(0) - B.get(i));
        for (int i = 1; i < size1; ++i)
        {
            for (int j = 1; j < size2; ++j)
            {
                if (1 == Config.DTWMethod)
                    distance[i][j] = Math.min(Math.min(distance[i-1][j], distance[i][j-1]), distance[i-1][j-1]) + Math.abs(A.get(i) - B.get(j));
                else if (2 == Config.DTWMethod)
                {
                    Double d1 = distance[i-1][j] + p;
                    Double d2 = distance[i][j-1] + p;
                    Double other = Math.abs(A.get(i) - B.get(j)) < t ? 0 : (Math.abs(A.get(i) - B.get(j)) - t);
                    Double d3 = distance[i-1][j-1] + other;
                    distance[i][j] = Math.min(d1, Math.min(d2, d3));
                }
                else if (3 == Config.DTWMethod)
                {
                    Double d1 = distance[i-1][j] + Math.abs(A.get(i) - B.get(j));
                    Double d2 = distance[i][j-1] + Math.abs(A.get(i) - B.get(j));
                    Double other = Math.abs(A.get(i) - B.get(j)) < t ? 0 : (Math.abs(A.get(i) - B.get(j) - t));
                    Double d3 = distance[i-1][j-1] + other;
                    distance[i][j] = Math.min(d1, Math.min(d2, d3));
                }
            }
        }

        return distance[size1-1][size2-1];
    }

    public ArrayList<Signature> refSigs;
    public Signature templateSig;
    public int templateIndex;
    public HashMap<String, Double> baseValueMap;
}

File: cloud/src/person.py
#coding:utf-8

import numpy

class Person(object):
    """
        Basic class of Person.
    """
    def __init__(self, config, refSigs, testSigs):
        """
            @param : config : configuration object
            @param : refSigs : reference signatures
            @param : testSigs : test signatures
        """

        self.config = config
        self.refSigs = refSigs
        self.testSigs = testSigs
        self.refCount = len(refSigs)
        self.templateSig = None
        # select template signature
        self.select_template()
        # calculate base distance
        self.calc_base_dis()

    def select_template(self):
        """
            Just select template signature.
            TODO: different signal with variant weight
        """
        def _size_norm(values):
            minValue = min(values)
            maxValue = max(values)
            _range = maxValue - minValue
            mV = 100.
            values = [mV * (v - minValue) / _range for v in values]
            return values

        self.config.logger.info("selecting template signature")
        refDis = []
        for i in range(self.refCount):
            dis = 0.0
            for j in range(self.refCount):
                if i == j:
                    continue
                comDisList = []
                totalWeight = 0.0
                for com in self.config.SigCompList:
                    signal1 = self.refSigs[i][com]
                    signal2 = self.refSigs[j][com]
                    signal1 = _size_norm(signal1)
                    signal2 = _size_norm(signal2)
                    comDisList.append(self.naive_dtw(signal1, signal2,
                        self.config.Penalization[com], self.config.Threshold[com]) * \
                                self.config.SignalWeight[com])
                    totalWeight += self.config.SignalWeight[com]
                dis += (sum(comDisList) / totalWeight)
            refDis.append(dis)

        self.templateIndex = refDis.index(min(refDis))
        self.config.logger.debug("template index : %d. RefSigDisList: %s" % (self.templateIndex, refDis))
        self.templateSig = self.refSigs[self.templateIndex]

    def calc_base_dis(self):
        """
        Calculate the base value in signal component of signatures
        """
        self.config.logger.info("Calculating base distance")
        self.base = {}

        for com in self.config.SigCompList:
            templateComList = []
            maxComList = []
            minComList = []
            avgComList = []
            medComList = []
            for i in range(self.refCount):
                if i == self.templateIndex:
                    continue
                comi = self.refSigs[i][com]
                templateComDis = self.naive_dtw(comi, self.templateSig[com],
                        self.config.Penalization[com], self.config.Threshold[com])
                templateComList.append(templateComDis)
                comDisList = []
                for j in range(self.refCount):
                    if i == j:
                        continue
                    comj = self.refSigs[j][com]
                    comDisList.append(self.naive_dtw(comi, comj,
                        self.config.Penalization[com], self.config.Threshold[com]))
                maxComList.append(max(comDisList))
                minComList.append(min(comDisList))
                avgComList.append(numpy.mean(comDisList))
                medComList.append(numpy.median(comDisList))
            if "template" in self.config.FeatureType[com]:
                self.base["template" + com] = numpy.mean(templateComList)
            if "max" in self.config.FeatureType[com]:
                self.base["max"+com] = numpy.mean(maxComList)
            if "min" in self.config.FeatureType[com]:
                self.base["min"+com] = numpy.mean(minComList)
            if "avg" in self.config.FeatureType[com]:
                self.base["avg"+com] = numpy.mean(avgComList)
            if "med" in self.config.FeatureType[com]:
                self.base["med"+com] = numpy.mean(medComList)
            self.config.logger.debug("Calculating signal: %s. %s" % (com,
                ", ".join(["%s:%s"%(items[0], items[1]) for items in self.base.items()])))

    def calc_dis(self, signature):
        """
            For given signature, calculate vector[] with normalization
        """
        featureVec = []
        for com in self.config.SigCompList:
            comSig = signature[com]
            comTem = self.templateSig[com]
            templateComDis = self.naive_dtw(comSig, comTem,
                    self.config.Penalization[com], self.config.Threshold[com])
            comDisList = []
            for i in range(self.refCount):
                comI = self.refSigs[i][com]
                dis = self.naive_dtw(comSig, comI,
                        self.config.Penalization[com], self.config.Threshold[com])
                comDisList.append(dis)
            maxComDis = max(comDisList)
            minComDis = min(comDisList)
            avgComDis = numpy.mean(comDisList)
            medComDis = numpy.median(comDisList)
            if "template" in self.config.FeatureType[com]:
                featureVec.append(templateComDis / self.base["template"+com])
            if "max" in self.config.FeatureType[com]:
                featureVec.append(maxComDis / self.base["max"+com])
            if "min" in self.config.FeatureType[com]:
                featureVec.append(minComDis / self.base["min"+com])
            if "avg" in self.config.FeatureType[com]:
                featureVec.append(avgComDis / self.base["avg"+com])
            if "med" in self.config.FeatureType[com]:
                featureVec.append(medComDis / self.base["med"+com])
            
        return featureVec

    def naive_dtw(self, A, B, p=5, t=5):
        penalization = p
        threshold = t
        len1 = len(A)
        len2 = len(B)
        distance = numpy.zeros((len1, len2))
        # initialize distance values
        distance[0][0] = abs(A[0] - B[0])
        for i in range(1, len1):
            distance[i][0] = distance[i-1][0] + abs(A[i] - B[0])

        for j in range(1, len2):
            distance[0][j] = distance[0][j-1] + abs(A[0] - B[j])

        for i in range(1, len1):
            for j in range(1, len2):
                if self.config.DTWMethod == 1:
                    distance[i][j] = min([distance[i-1][j], distance[i][j-1],
                            distance[i-1][j-1]]) + abs(A[i]-B[j])
                elif self.config.DTWMethod == 2:
                    # DTWMethod 2
                    d1 = distance[i-1][j] + penalization
                    d2 = distance[i][j-1] + penalization
                    other = 0 if (abs(A[i] - B[j]) < threshold) else (abs(A[i] - B[j]) - threshold)
                    d3 = distance[i-1][j-1] + other
                    distance[i][j] = min([d1, d2, d3])
                elif self.config.DTWMethod == 3:
                    d1 = distance[i-1][j] + abs(A[i] - B[j])
                    d2 = distance[i][j-1] + abs(A[i] - B[j])
                    other = 0 if (abs(A[i] - B[j]) < threshold) else (abs(A[i] - B[j]) - threshold)
                    d3 = distance[i-1][j-1] + other
                    distance[i][j] = min([d1, d2, d3])
                    
        return distance[len1-1][len2-1]

class PersonTest(Person):
    def __init__(self, config, refSigs):
        super(PersonTest, self).__init__(config, refSigs, None)
